src/sys/options.cxx

```diff
@@ -562,33 +643,100 @@ bool Options::operator<(const char* other) const {
+  // Check if the option should count as having been used due to its source
+  const auto has_excluded_source = [&exclude_sources](const Options& option) -> bool {
+    if (not option.hasAttribute("source")) {
+      return false;
     }
-  }
-  if (allused) {
-    output_info << _("All options used\n");
-  } else {
-    output_info << _("Unused options:\n");
-    for (const auto &it : children) {
-      if (it.second.is_value && !it.second.value_used) {
-        output_info << "\t" << full_name << ":" << it.first << " = "
-                    << bout::utils::variantToString(it.second.value);
-        if (it.second.attributes.count("source"))
-          output_info << " (" << bout::utils::variantToString(it.second.attributes.at("source")) << ")";
-        output_info << endl;
+    const auto source = option.attributes.at("source").as<std::string>();
+    return std::find(exclude_sources.begin(), exclude_sources.end(), source)
+           != exclude_sources.end();
+  };
+
+  const auto conditionally_used = [](const Options& option) -> bool {
+    if (not option.hasAttribute(conditionally_used_attribute)) {
+      return false;
+    }
+    return option.attributes.at(conditionally_used_attribute).as<bool>();
+  };
+
+  // Copy this object, and then we're going to chuck out everything
+  // that has been used. This turns out to be easier than copying just
+  // the unused options into an empty instance
+  Options unused = *this;
+
+  if (unused.is_value) {
+    // If this is from an excluded source, count it as being used
+    if (has_excluded_source(unused) or conditionally_used(unused)) {
+      unused.value_used = true;
+    }
+    // We don't have a nice way to "clear" the value, so if it was
+    // used, mark it as no longer a value: if it has been used, this
+    // does nothing
+    unused.is_value = not unused.value_used;
+    return unused;
+  }
+
+  // This loop modifies the map in the loop, so we need to manually
+  // manage the iterator
+  for (auto child = unused.children.begin(); child != unused.children.end();) {
+    // Remove the child if it's been used or if it's from a source we
+    // should count as having been used
+    if (child->second.is_value
+        and (child->second.value_used or has_excluded_source(child->second)
+             or conditionally_used(child->second))) {
+      child = unused.children.erase(child);
+      continue;
+    }
+
+    if (child->second.is_section) {
+      // Recurse down and replace this section by its "unused" version
+      child->second = child->second.getUnused();
+      // If all of its children have been used, then we can remove it
+      // as well
+      if (child->second.children.empty()) {
+        child = unused.children.erase(child);
+        continue;
       }
     }
-  }
-  for (const auto &it : children) {
-    if (it.second.is_section) {
-      it.second.printUnused();
+
+    // What is it doing here?!
```

Comment:
This is covers the case where the `Options` is neither a value nor a section. I didn't quite pin down exactly when that can happen, but apparently it does. I'll try digging into it again or at least make the comment better!

Reply:
It turns out this can happen when calling `option.getSection()` with a section that doesn't exist. This calls `operator[]` which creates a new `Options`, with default values for `is_section` and `is_value` (both `false`).

A possible fix is:

```cpp
Options* Options::getSection(const std::string& name) {
  auto* section = &(*this)[name];
  section->is_section = true;
  return section;
}
```

but doing something like

```cpp
Options section = Options::root()["new section"];
```

still won't work.

Another solution would be to remove `is_section` and just have `is_value`. A new `Options` would default to being a section, but assignment would turn it into a value.

Having both let's us do things like:

```ini
name = value

[name]
stuff = foo
```

but do we really want to support having both a section and a value with the same name at the same level?

@bendudson Is there something else I'm missing that requires having both `is_section` and `is_value`?

Reply:
I hoped to get rid of this too,  but at the moment we have a top level option `restart` and a section of the same name